src/query/subquery_cache.c

```diff
@@ -0,0 +1,749 @@
+	}
+    }
+}
+
+/*
+ * sq_cache_destroy () - Destroys the cache for a given XASL node.
+ *   xasl(in): The XASL node for which the cache is being destroyed.
+ *
+ * This function destroys the cache associated with a given XASL node. It clears all cache entries and then destroys the hash
+ * table itself. It also resets cache-related flags and counters for the XASL node. This function is called when a XASL node is
+ * no longer needed or before it is deallocated.
+ */
+void
+sq_cache_destroy (xasl_node * xasl)
+{
+  if (xasl->sq_cache_flag & SQ_CACHE_INITIALIZED_FLAG)
+    {
+      er_log_debug (ARG_FILE_LINE,
+		    "destroy sq_cache at xasl %p\ncache info : \n\thit : %10lu\n\tmiss: %10lu\n\tsize: %10lu Bytes\n",
+		    xasl, xasl->sq_cache_hit, xasl->sq_cache_miss, xasl->sq_cache_size);
+      sq_cache_drop_all (xasl);
+      mht_destroy (xasl->sq_cache_ht);
+    }
+  xasl->sq_cache_flag = 0;
+  xasl->sq_cache_ht = NULL;
+}
+
+/*
+ * execute_regu_variable_xasl_with_sq_cache () - Executes a regu variable XASL with support for scalar subquery result caching.
+ *   return: False if execution should proceed after caching logic, True otherwise.
+ *   thread_p(in): Thread context.
+ *   regu_var(in): The regu variable to be executed.
+ *   vd(in): Value descriptor for parameter values.
+ *
+ * This function attempts to execute a regu variable's XASL with caching logic for scalar subquery results. It checks if the
+ * XASL node associated with the regu variable is eligible for caching and either retrieves the cached result or executes the
+ * XASL and caches the result if appropriate. This function aims to improve performance by avoiding redundant executions of
+ * scalar subqueries.
+ */
+int
+execute_regu_variable_xasl_with_sq_cache (THREAD_ENTRY * thread_p, REGU_VARIABLE * regu_var, void *vd)
+{
+  if (regu_var->xasl && !(regu_var->xasl->sq_cache_flag & SQ_CACHE_NOT_CACHING_FLAG)
+      && (regu_var->xasl->status == XASL_CLEARED || regu_var->xasl->status == XASL_INITIALIZED))
+    {
+
+      if (regu_var->xasl->sq_cache_flag & SQ_CACHE_ENABLED_FLAG)
+	{
+	  if (sq_get (regu_var->xasl, regu_var))
+	    {
+	      regu_var->xasl->status = XASL_SUCCESS;
+	      return true;
+	    }
+	  EXECUTE_REGU_VARIABLE_XASL (thread_p, regu_var, (val_descr *) vd);
+	  if (CHECK_REGU_VARIABLE_XASL_STATUS (regu_var) != XASL_SUCCESS)
+	    {
+	      return false;
+	    }
+	  sq_put (regu_var->xasl, regu_var);
+
+	  return false;
+	}
+
+      else if (regu_var->xasl->sq_cache_flag == 0)
+	{
+	  regu_var->xasl->sq_cache_flag |= SQ_CACHE_ENABLED_FLAG;
+	  if (!(regu_var->xasl->sq_cache_flag & SQ_CACHE_NOT_CACHING_CHECKED_FLAG))
+	    {
+	      if (sq_walk_xasl_check_not_caching (regu_var->xasl))
+		{
+		  regu_var->xasl->sq_cache_flag |= SQ_CACHE_NOT_CACHING_FLAG;
+		}
+	      regu_var->xasl->sq_cache_flag |= SQ_CACHE_NOT_CACHING_CHECKED_FLAG;
+	    }
+	}
+    }
+
+  EXECUTE_REGU_VARIABLE_XASL (thread_p, regu_var, (val_descr *) vd);
```

Comment:
I think it is better for readability to separate execute_regu_variable_xasl_with_sq_cache and EXECUTE_REGU_VARIABLE_XASL.
How about keeping EXECUTE_REGU_VARIABLE_XASL in its original location?

Reply:
I feel that among the functions in `src/query/subquery_cache.c`, the `execute_regu_variable_xasl_with_sq_cache` function is something different.
I think only functions for managing the result cache of subqueries should be in `src/query/subquery_cache.c`.
It is strange that the function in `src/query/subquery_cache.c` directly calls the `qexec_execute_mainblock` function.

Reply:
I don't think it's strange to call the sq_cache function before calling the qexec_execute_mainblock function in fetch_peek_dbval.
If qexec_execute_mainblock was executed, there was no result cache in sq_cache, and we call the sq_cache function again to cache the result.

Reply:
I wish the if statement was simple.
```
case TYPE_CONSTANT: /* fetch constant-column value */
  /* is not constant */
  REGU_VARIABLE_SET_FLAG (regu_var, REGU_VARIABLE_FETCH_NOT_CONST);
  assert (!REGU_VARIABLE_IS_FLAGED (regu_var, REGU_VARIABLE_FETCH_ALL_CONST));
  
  if (sq_check_enable (...) == ENABLE) /* pseudocode */
    {
      if (sq_find_cache (...) == NOT_FOUND) /* pseudocode */
        {
          /* execute linked query */
          EXECUTE_REGU_VARIABLE_XASL (thread_p, regu_var, vd);
          if (CHECK_REGU_VARIABLE_XASL_STATUS (regu_var) != XASL_SUCCESS)
            {
              goto exit_on_error;
            }
        
           /* pseudocode */
          if (sq_cache (...) != SUCCESS)
            {
              sq_disable_cache (...);
            }
        }
      else
        {
          /* FOUND */
          regu_var->xasl->status = XASL_SUCCESS;
        }
    }
  else
    {
      /* execute linked query */
      EXECUTE_REGU_VARIABLE_XASL (thread_p, regu_var, vd);
      if (CHECK_REGU_VARIABLE_XASL_STATUS (regu_var) != XASL_SUCCESS)
        {
        goto exit_on_error;
        }
    }
        
  *peek_dbval = regu_var->value.dbvalptr;
  break;
```

Reply:
I think it would be good to create the `EXECUTE_REGU_VARIABLE_XASL_WITH_SQ_CACHE` macro function for this code.
However, the location must be `xasl.h`, not `sq_cache`.